### PR TITLE
Improved templates APIs

### PR DESCRIPTION
**Pull request by mishushakov**
- Added fields to API:
  - createdAt
  - updatedAt
  - createdBy
  - lastSpawnedAt
  - spawnCount
  - buildCount
- Added PATCH method to /templates/{templateID} for updating template info in db
  - public
- Added createdBy field to envs DB - migration required

<details open>
<summary>:sparkles: <i><h3>Description by Callstackai</h3></i></summary>

This PR improves the templates APIs by adding new fields to the API, implementing a PATCH method for updating template information, and modifying the database schema to include a 'createdBy' field.

<details>
<summary><b>Diagrams of code changes</b></summary>

```mermaid
sequenceDiagram
    Client->>+API: PATCH /templates/{templateID}
    Note over API: New endpoint to update template visibility
    
    API->>+DB: Get template & validate access
    DB-->>-API: Template data
    
    API->>+DB: Update template.public field
    DB-->>-API: Update confirmed
    
    Note over DB: New template fields added:
    Note over DB: - created_by (User)
    Note over DB: - created_at
    Note over DB: - updated_at  
    Note over DB: - last_spawned_at
    Note over DB: - spawn_count
    Note over DB: - build_count
    
    API-->>-Client: 200 OK
```

</details>

<details>
<summary><b>Files Changed</b></summary>
<table>
<tr><th>File</th><th>Summary</th></tr>
<tr><td><a href=https://github.com/e2b-dev/infra/pull/206/files#diff-bfa6ea3da4f29a1db196d3d485417366fd4bc3e73662eb6c011ae4a236246925>packages/api/internal/api/api.gen.go</a></td><td>Added PATCH method to the ServerInterface for updating template information.</td></tr>
<tr><td><a href=https://github.com/e2b-dev/infra/pull/206/files#diff-ae30b8e75db688f70c3d6d32afb639f3a10ef77f0ba75be27b5e440eb338615c>packages/api/internal/api/spec.gen.go</a></td><td>Updated OpenAPI specification to include new fields in the template schema.</td></tr>
<tr><td><a href=https://github.com/e2b-dev/infra/pull/206/files#diff-a09a97c0b315e6b8c0fd59031fb60e270d5650cd9294954e6e2fface867c2bc7>packages/api/internal/api/types.gen.go</a></td><td>Defined new types for the additional fields in the template.</td></tr>
<tr><td><a href=https://github.com/e2b-dev/infra/pull/206/files#diff-2bc872d7c797eb821ef850762530ecbbd9dec361ec80a7813117a73681f1658e>packages/shared/migrations/20241127174604.sql</a></td><td>Added migration to include 'created_by' field in the envs table.</td></tr>
<tr><td><a href=https://github.com/e2b-dev/infra/pull/206/files#diff-264b91e96c33a99c5a2a78ffe2c06a441a10f2cc370b28e7f1667291fb625f2c>packages/shared/pkg/models/env.go</a></td><td>Updated Env model to include new fields.</td></tr>
<tr><td><a href=https://github.com/e2b-dev/infra/pull/206/files#diff-7da427fdebe5555e63b6e24600e140a8139ea842b05413c84b0101169dfc4979>packages/shared/pkg/models/user.go</a></td><td>Updated User model to include created environments.</td></tr>
<tr><td><a href=https://github.com/e2b-dev/infra/pull/206/files#diff-5b7a50ab5c4be5545227c7c591b86773a13143499d6d2f101df67776abdb847d>spec/openapi.yml</a></td><td>Updated OpenAPI definitions to reflect new fields and endpoints.</td></tr>

</table>
</details>

*This PR includes files in programming languages that we currently do not support. We have not reviewed files with the extensions `.sql`, `.sum`, `.yml`. <a href=https://docs.callstack.ai/introduction>See list of supported languages.</a>*

</details>

### CODE REVIEW

File: packages/shared/pkg/models/user.go [hunk 1]
```diff
@@ -49,10 +51,19 @@ func (e UserEdges) TeamsOrErr() ([]*Team, error) {
...
+		return e.CreatedEnvs, nil
+	}
+	return nil, &NotLoadedError{edge: "created_envs"}
+}
+
 // AccessTokensOrErr returns the AccessTokens value or an error if the edge
 // was not loaded in eager-loading.
 func (e UserEdges) AccessTokensOrErr() ([]*AccessToken, error) {
-	if e.loadedTypes[1] {
+	if e.loadedTypes[2] {
```
[the remark this answers is not in the document]
Comment (mishushakov) [reply packages/shared/pkg/models/user.go:66]: Mr. AI, I did not write this code

File: packages/api/internal/handlers/template_request_build.go [hunk 2]
```diff
@@ -203,6 +203,7 @@ func (a *APIStore) TemplateRequestBuild(c *gin.Context, templateID api.TemplateI
 		Create().
 		SetID(templateID).
 		SetTeamID(team.ID).
+		SetCreatedBy(*userID).
```
[the remark this answers is not in the document]
Comment (mishushakov) [reply packages/api/internal/handlers/template_request_build.go:206]: Mr. AI, userID will always be present as long as you provide API key to the request

File: packages/api/internal/handlers/template_update.go [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+	}
+
+	cleanedAliasOrEnvID, err := id.CleanEnvID(aliasOrTemplateID)
+	if err != nil {
+		a.sendAPIStoreError(c, http.StatusBadRequest, fmt.Sprintf("Invalid env ID: %s", aliasOrTemplateID))
+
+		err = fmt.Errorf("invalid env ID: %w", err)
+		telemetry.ReportCriticalError(ctx, err)
+
+		return
```
[the remark this answers is not in the document]
Comment (mishushakov) [reply packages/api/internal/handlers/template_update.go:38]: Mr. AI, this part was copied from template_delete.go

File: packages/shared/pkg/models/env.go [hunk 4]
```diff
@@ -137,6 +157,13 @@ func (e *Env) assignValues(columns []string, values []any) error {
 			} else if value != nil {
 				e.TeamID = *value
 			}
+		case env.FieldCreatedBy:
+			if value, ok := values[i].(*sql.NullScanner); !ok {
+				return fmt.Errorf("unexpected type %T for field created_by", values[i])
+			} else if value.Valid {
+				e.CreatedBy = new(uuid.UUID)
+				*e.CreatedBy = *value.S.(*uuid.UUID)
+			}
```
[the remark this answers is not in the document]
Comment (mishushakov) [reply packages/shared/pkg/models/env.go:166]: Mr. AI, I did not write this code

File: packages/api/internal/handlers/template_update.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (jakubno) [reply packages/api/internal/handlers/template_update.go:38]: But he's right, the naming is mess. There are some old naming as instance, session for sandbox and env for template...

File: packages/api/internal/handlers/template_update.go [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+	}
+
+	// Prepare info for rebuilding env
```
Comment (jakubno) [inline packages/api/internal/handlers/template_update.go]:
```suggestion
```

**Pull request merged**